File: Roadmap/02 - FUNCIONES Y ALCANCE/go/raynerpv2022.go
package main

import "fmt"

func func_sin_p() {
	fmt.Println("hoy es   6 de agosto")
}

func func_1_p(mes string) {
	fmt.Printf("hoy es  6 de %s \n", mes)
}

func func_2_p(dia int, mes string) {
	fmt.Printf("hoy es %d de %s \n", dia, mes)
}

// funcion variatica para valores por defecto
func func_value_default(fecha ...string) {
	dia := "26"
	mes := "enero"
	if len(fecha) > 0 {
		dia = fecha[0]
		mes = fecha[1]
	}
	fmt.Printf("hoy es  %s de %s \n", dia, mes)
}

func returnMult(dia, mes string) (string, string) {
	return dia, mes
}

func Mult3y5(tres, cinco string) int {
	number := 0
	for i := 1; i <= 100; i++ {
		if i%3 == 0 && i%5 == 0 {
			fmt.Println(tres, cinco)
		} else if i%3 == 0 {
			fmt.Println(tres)
		} else if i%5 == 0 {
			fmt.Println(cinco)
		} else {
			number++
			fmt.Println(i)
		}
	}
	return number

}

func func_var_param(names ...string) {
	long := len(names)
	if long > 0 {
		fmt.Printf("existen %d parametros\n", long)
		for _, fecha := range names {
			fmt.Println(fecha)
		}
	} else {
		fmt.Println("no existen parametros")

	}
}

func main() {
	fmt.Println(Mult3y5("TRES", "CINCO"))
	func_1_p("noviembre")
	func_2_p(3, "octubre")
	func_sin_p()
	func_value_default()
	func_value_default("31", "diciembre")
	dia, mes := returnMult("1", "enero")
	fmt.Println(dia, mes)
	func_var_param()
	func_var_param("31", "mayo")

}

File: Roadmap/02 - FUNCIONES Y ALCANCE/python/raynerpv2022.py
"""
 * EJERCICIO:
 * - Crea ejemplos de funciones básicas que representen las diferentes
 *   posibilidades del lenguaje:
 *   Sin parámetros ni retorno, con uno o varios parámetros, con retorno...
 * - Comprueba si puedes crear funciones dentro de funciones.
 * - Utiliza algún ejemplo de funciones ya creadas en el lenguaje.
 * - Pon a prueba el concepto de variable LOCAL y GLOBAL.
 * - Debes hacer print por consola del resultado de todos los ejemplos.
 *   (y tener en cuenta que cada lenguaje puede poseer más o menos posibilidades)
 *
 * DIFICULTAD EXTRA (opcional):
 * Crea una función que reciba dos parámetros de tipo cadena de texto y retorne un número.
 * - La función imprime todos los números del 1 al 100. Teniendo en cuenta que:
 *   - Si el número es múltiplo de 3, muestra la cadena de texto del primer parámetro.
 *   - Si el número es múltiplo de 5, muestra la cadena de texto del segundo parámetro.
 *   - Si el número es múltiplo de 3 y de 5, muestra las dos cadenas de texto concatenadas.
 *   - La función retorna el número de veces que se ha impreso el número en lugar de los textos.
 *
 * Presta especial atención a la sintaxis que debes utilizar en cada uno de los casos.
 * Cada lenguaje sigue una convenciones que debes de respetar para que el código se entienda.
"""

def func_sin_p():
  print("hoy es   6 de agosto")

def func_1_p(mes):
  print(f"hoy es  6 de {mes}")

def func_2_p(dia, mes):
  print(f"hoy es  {dia} de {mes}")

def func_value_default(dia = 26,mes="enero"):
  print(f"hoy es  {dia} de {mes}")

def func_value_dmult(dia = 26,mes="enero"):
  return dia, mes

def func_par_var(*var):
  
  if len(var) >0:
    print(f" existen {len(var)} parametros")
    for i in var:
        print(i)
  else :
    print("No existen Parametros")

def func_k_v(** data):
  for k,v in data.items():
    print(f"key : {k}, value : {v}")




#  Dificultad Extra
def print_number(cad1, cad2):
 number = 0
 for i in range(1, 101):
   
   if i % 3 ==0 and i % 5 ==0:
     print(cad1+" "+cad2)
   elif i % 3 == 0 :
      print(cad1)
   elif i % 5 == 0:
     print(cad2)
   else:
     number+=1
     print(i)
 return number

print(print_number("TRES","CINCO"))
func_sin_p()
func_1_p("enero")
func_2_p(6,"diciembre")
func_value_default(22,"diciembre")
func_value_default()
dia = func_value_dmult()[0]
mes = func_value_dmult()[1]
print(dia, mes)
func_par_var(1,2,3,4,5)
func_par_var()
func_k_v(a=1,b=2,c=3)

# local y global variables
global_var = "ray"
print("Global ", global_var)

def local_Function():
  global_var = "21"
  print("LOcal variable", global_var)
local_Function()
print("Global ", global_var)

